### PR TITLE
test: single asset low-level inverse calculations (TestCalcSingleAssetInAndOut_InverseRelationship)

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This PR is focused on improving code and logic test coverage in the balancer pool. In addition, it helps to continue investigating rounding errors in https://github.com/osmosis-labs/osmosis/issues/1196 and https://github.com/osmosis-labs/osmosis/pull/1307 by improving test coverage across low level Pool methods

Concretely, this PR adds inverse tests for `calcPoolSharesOutGivenSingleAssetIn` and `calcSingleAssetInGivenPoolSharesOut` with zero and non-zero swap fees.

Please note that the commented-out test is failing. It might be due to the low pool balance. I'm not sure how big of an issue this is

## Brief change log

- port `calcSingleAssetInGivenPoolSharesOut` from `v7.x`
   * this method is to be used in #1307
- add inverse table tests with swap fees

## Testing and Verifying

This change added tests and can be verified as follows:
```
go test -timeout 30s -run ^TestCalcSingleAssetInAndOut_InverseRelationship$ github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/balancer
```

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/gamm/pool-models/balancer/amm_test.go [hunk 1]
```diff
@@ -186,3 +187,114 @@ func TestCalculateAmountOutAndIn_InverseRelationship_ZeroSwapFee(t *testing.T) {
...
+			initialWeightOut: 148,
+			initialWeightIn:  57,
+		},
+		{
+			initialPoolOut:   1_000,
+			tokenOut:         25,
+			initialWeightOut: 100,
+			initialWeightIn:  100,
+		},
+		// {
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm_test.go:244]: This test is failing. Question to reviewers - is this test invalid due to the low pool balance compared to tokenOut or this is a bug?

I have a feeling that it's the former but wanted to double-check

File: x/gamm/pool-models/balancer/amm.go [hunk 2]
```diff
@@ -225,6 +229,23 @@ func calcPoolSharesOutGivenSingleAssetIn(
 	return poolAmountOut
 }
 
+// calcSingleAssetInGivenPoolSharesOut returns token amount in with fee included
+// given the swapped out shares amount, using solveConstantFunctionInvariant
+func calcSingleAssetInGivenPoolSharesOut(
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm.go:234]: This function will be used in #1307

File: x/gamm/pool-models/balancer/amm_test.go (hunk 1 above)
Comment (ValarDragon) [reply x/gamm/pool-models/balancer/amm_test.go:244]: That sounds right, as long as this is off by one

File: x/gamm/pool-models/balancer/export_test.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package balancer
+
+var (
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/export_test.go:3]: Why are we aliasing here making private one public?

**Review by alexanderbez — APPROVED** [review]

File: x/gamm/pool-models/balancer/export_test.go (hunk 3 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/export_test.go:3]: To be able to test unexported functions in tests. Files with `_test.go` do not get compiled into the final binary. As a result, these aliases are only accessible in tests

File: x/gamm/pool-models/balancer/amm_test.go (hunk 1 above)
Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm_test.go:259]: Created an issue to uncomment: https://github.com/osmosis-labs/osmosis/issues/1359

**Pull request merged**